src/graphics/Window.zig

```diff
@@ -1,25 +1,381 @@
+		return self.gamepadState.count() > 0;
+	}
+	pub fn controllerMappingsDownloading(self: *Gamepad) bool {
+		return !self.*.controllerMappingsDownloaded;
+	}
+	const ControllerMappingDownloadTask = struct { // MARK: ControllerMappingDownloadTask
+		curTimestamp: i128,
+		controllerMappingsDownloaded: *bool,
+		const vtable = main.utils.ThreadPool.VTable{
+			.getPriority = @ptrCast(&getPriority),
+			.isStillNeeded = @ptrCast(&isStillNeeded),
+			.run = @ptrCast(&run),
+			.clean = @ptrCast(&clean),
+		};
+
+		pub fn schedule(curTimestamp: i128, controllerMappingsDownloaded: *bool) void {
+			const task = main.globalAllocator.create(ControllerMappingDownloadTask);
+			task.* = ControllerMappingDownloadTask {
+				.curTimestamp = curTimestamp,
+				.controllerMappingsDownloaded = controllerMappingsDownloaded,
+			};
+			main.threadPool.addTask(task, &vtable);
+		}
+
+		pub fn getPriority(_: *ControllerMappingDownloadTask) f32 {
+			return 64.0;
+		}
+
+		pub fn isStillNeeded(self: *ControllerMappingDownloadTask, _: i64) bool {
+			return !@atomicLoad(bool, self.*.controllerMappingsDownloaded, std.builtin.AtomicOrder.acquire);
+		}
+
+		pub fn run(self: *ControllerMappingDownloadTask) void {
+			defer self.clean();
+			var client: std.http.Client = .{.allocator = main.stackAllocator.allocator};
+			var list = std.ArrayList(u8).init(main.stackAllocator.allocator);
+			defer client.deinit();
+			defer list.deinit();
+			defer @atomicStore(bool, self.*.controllerMappingsDownloaded, true, std.builtin.AtomicOrder.release);
+			_ = client.fetch(.{
+				.method = .GET,
+				.location = .{.url = "https://raw.githubusercontent.com/mdqinc/SDL_GameControllerDB/master/gamecontrollerdb.txt"},
+				.response_storage = .{ .dynamic = &list }
+			}) catch |err| {
+				std.log.err("Failed to download controller mappings: {s}", .{@errorName(err)});
+				return;
+			};
+			files.write("gamecontrollerdb.txt", list.items) catch |err| {
+				std.log.err("Failed to write controller mappings: {s}", .{@errorName(err)});
+				return;
+			};
+			const timeStampStr = std.fmt.allocPrint(main.stackAllocator.allocator, "{d}", .{self.*.curTimestamp}) catch unreachable;
+			defer main.stackAllocator.free(timeStampStr);
+			files.write("gamecontrollerdb.stamp", timeStampStr) catch |err| {
+				std.log.err("Failed to write controller mappings: {s}", .{@errorName(err)});
+				return;
+			};
+		}
 
+		pub fn clean(self: *ControllerMappingDownloadTask) void {
+			main.globalAllocator.destroy(self);
+		}
+	};
+	pub fn downloadControllerMappings(self: *Gamepad) void {
+		var needDownload: bool = false;
+		const curTimestamp = std.time.nanoTimestamp();
+		if (settings.downloadControllerMappings) {
+			var timestamp: i128 = 0;
+			if (files.openDir(".") catch null) |dir| {
+				var _dir: files.Dir =dir;
+				if (_dir.hasFile("gamecontrollerdb.stamp")) {
+					const stamp = _dir.read(main.globalAllocator, "gamecontrollerdb.stamp") catch unreachable;
```

Comment:
This error is not unreachable! The file might for example have incorrect permissions.
also: stackAllocator